core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,329 @@
+  object CacheSettings {
+    val default: CacheSettings = CacheSettings(0.seconds, CacheScope.Public)
+  }
+
+  /**
+   * Overrides the cache hint for a particular field. This can be used to dynamically set the cache hint within
+   * a resolver
+   */
+  def setCacheHint(hint: CacheHint): UIO[Unit] = cacheOverride.update {
+    case Some(hint0) => Some(mostRestrictive(hint0, hint))
+    case None        => Some(hint)
+  }
+
+  /**
+   * Disables caching for the current query
+   */
+  def disableCaching: UIO[Unit] = cacheOverride.set(Some(CacheHint(Some(0.seconds))))
+
+  case class CachePolicy(hint: CacheHint) {
+    def merge(that: CachePolicy): CachePolicy =
+      CachePolicy(mostRestrictive(hint, that.hint))
+
+    def restrict(restricted: Option[CacheHint]): CachePolicy = restricted match {
+      case Some(h) => copy(hint = mostRestrictive(hint, h))
+      case None    => this
+    }
+
+    def toResponseValue(toHeader: CachePolicy => String) =
+      ObjectValue(
+        List(
+          "version"    -> IntValue(2),
+          "httpHeader" -> StringValue(toHeader(this))
+        )
+      )
+  }
+
+  sealed trait CacheScope
+
+  object CacheScope {
+
+    @GQLName("PRIVATE")
+    case object Private extends CacheScope {
+      override def toString: String = "PRIVATE"
+    }
+
+    @GQLName("PUBLIC")
+    case object Public extends CacheScope {
+      override def toString: String = "PUBLIC"
+    }
+
+    implicit val ord: Ord[CacheScope] = Ord.make {
+      case (Private, Private) => Ordering.Equals
+      case (Public, Public)   => Ordering.Equals
+      case (Private, Public)  => Ordering.LessThan
+      case (Public, Private)  => Ordering.GreaterThan
+    }
+
+    implicit val ordering: scala.math.Ordering[CacheScope] = ord.toScala
+
+    val _type = __Type(
+      __TypeKind.ENUM,
+      Some("CacheScope"),
+      Some("The scope of the cache"),
+      enumValues = _ =>
+        Some(
+          List(
+            __EnumValue(
+              "PRIVATE",
+              Some("The cache is private to the user"),
+              deprecationReason = None,
+              isDeprecated = false,
+              directives = None
+            ),
+            __EnumValue(
+              "PUBLIC",
+              Some("The cache is public"),
+              deprecationReason = None,
+              isDeprecated = false,
+              directives = None
+            )
+          )
+        )
+    )
+  }
+
+  case class CacheHint(
+    maxAge: Option[Duration] = None,
+    scope: Option[CacheScope] = None,
+    inheritMaxAge: Boolean = false
+  ) { self =>
+    def toHeaderString: String = {
+      val computedMaxAge = maxAge.fold(0L)(_.toSeconds)
+      val cacheable      = computedMaxAge > 0L
+      if (!cacheable) "no-store"
+      else {
+        val b = new StringBuilder()
+        b.append("max-age=")
+        b.append(computedMaxAge)
+        scope.foreach { s =>
+          b.append(", ")
+          b.append(s.toString.toLowerCase)
+        }
```

Comment:
By default, the `public` scope is used. Should we default to `private` instead?

> If you forget to add private to a response with personalized content, then that response can be stored in a shared cache and end up being reused for multiple users, which can cause personal information to leak.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

Reply:
Not sure, I think the default from Apollo is public, and likewise it seems that's the case for the http header as well. Nonetheless, this already deviates somewhat so we could define our own semantic if we want.

Reply:
Yeah after thinking it through a bit more, I think it might be better to keep behaviour as consistent with the standard semantics.

Perhaps we should add it in the scaladoc somewhere & release notes that if no scope is defined, `PUBLIC` is used by default. The `ApolloCaching` wrapper didn't use headers and it might cause some issues for users that migrate to this wrapper who forgot to set a scope